File: backend/endpoints/documents.py
"""Routes for updating the database.

Every route in this file is access controlled.
"""

from __future__ import annotations
import asyncio
from enum import StrEnum
from typing import Iterator
import flask

from backend.common import connect
from backend.common.backend_exceptions import HandledException
from backend.common.database import DocumentRef, DocumentsRef, LibraryRef
from backend.common.app_access import require_access_level
from backend.common.firebase_storage import upload_thumbnails
from backend.common.models import (
    Element,
)
from backend.common.models import Document
from backend.common.vendors import parse_vendors
from backend.endpoints.configurations import parse_onshape_configuration
from backend.common.reload_context import (
    ReloadContext,
)
from onshape_api.api.api_base import Api
from onshape_api.endpoints import documents
from onshape_api.endpoints.configurations import get_configuration
from onshape_api.endpoints.documents import ElementType, get_document
from onshape_api.endpoints.versions import get_latest_version_path
from onshape_api.paths.doc_path import (
    DocumentPath,
    ElementPath,
    InstancePath,
)

router = flask.Blueprint("documents", __name__)


def save_element(
    api: Api,
    document_ref: DocumentRef,
    version_path: InstancePath,
    onshape_element: dict,
    reload_context: ReloadContext,
) -> str:
    """
    Parameters:
        element: A part studio or assembly returned by the /elements endpoint.
    """

    element_type: ElementType = onshape_element["elementType"]
    element_name = onshape_element["name"]  # Use the name of the tab
    element_id = onshape_element["id"]
    microversion_id = onshape_element["microversionId"]

    path = ElementPath.from_path(version_path, element_id)

    onshape_configuration = get_configuration(api, path)
    configuration_id = None
    configuration = None
    if len(onshape_configuration["configurationParameters"]) > 0:
        configuration = parse_onshape_configuration(onshape_configuration)
        # Re-use element db id since configurations can't be shared
        document_ref.configurations.configuration(element_id).set(configuration)
        configuration_id = element_id

    thumbnailUrls = upload_thumbnails(api, path, microversion_id)

    preserved_element = reload_context.get_element(element_id)
    document_ref.elements.element(element_id).set(
        Element(
            name=element_name,
            vendors=parse_vendors(element_name, configuration),
            elementType=element_type,
            documentId=version_path.document_id,
            instanceId=version_path.instance_id,
            microversionId=microversion_id,
            configurationId=configuration_id,
            isVisible=preserved_element.isVisible,
            isOpenComposite=preserved_element.isOpenComposite,
            fastenInfo=preserved_element.fastenInfo,
            thumbnailUrls=thumbnailUrls,
        ),
    )
    return element_id


class EntryType(StrEnum):
    GROUP = "BTElementGroup-1458"
    ELEMENT = "BTDocumentElementReference-2484"


def get_ordered_element_ids(contents: dict) -> Iterator[str]:
    """Returns a list of element_ids in the same order as they are in an Onshape document."""
    for entry in contents["folders"]["groups"]:
        yield from traverse_entry(entry)


def traverse_entry(entry: dict) -> Iterator[str]:
    entry_type = entry["btType"]
    if entry_type == EntryType.GROUP:
        for entry in entry["groups"]:
            yield from traverse_entry(entry)
    elif entry_type == EntryType.ELEMENT:
        yield entry["elementId"]


def get_valid_elements(contents: dict) -> Iterator[dict]:
    for onshape_element in contents["elements"]:
        if onshape_element["elementType"] not in [
            ElementType.ASSEMBLY,
            ElementType.PART_STUDIO,
        ]:
            continue

        yield onshape_element


def get_elements_to_reload(
    document_ref: DocumentRef,
    valid_elements: list[dict],
    reload_context: ReloadContext,
) -> Iterator[dict]:
    for onshape_element in valid_elements:
        element = document_ref.elements.element(onshape_element["id"]).maybe_get()
        if element == None:
            yield onshape_element
            continue

        if reload_context.should_reload_element(
            onshape_element["id"], onshape_element["microversionId"]
        ):
            yield onshape_element


def get_element_microversion_id(contents: dict, element_id: str) -> str | None:
    for onshape_element in contents["elements"]:
        if onshape_element["id"] == element_id:
            return onshape_element["microversionId"]
    return None


async def save_document(
    api: Api,
    document_ref: DocumentRef,
    version_path: InstancePath,
    reload_context: ReloadContext,
) -> int:
    """Loads all of the elements of a given document into the database.

    Note this function does NOT update documentOrder; it is up to the caller to add it themselves.
    """
    document_id = version_path.document_id

    # Get document synchronously up front to do validation
    onshape_document = await asyncio.to_thread(
        documents.get_document, api, version_path
    )
    thumbnail_element_id = onshape_document["documentThumbnailElementId"]
    if thumbnail_element_id == "":
        raise HandledException(
            onshape_document["name"] + " does not have a thumbnail tab set"
        )

    contents = await asyncio.to_thread(documents.get_contents, api, version_path)

    thumbnail_path = ElementPath.from_path(version_path, thumbnail_element_id)
    thumbnail_microversion_id = get_element_microversion_id(
        contents, thumbnail_element_id
    )

    if thumbnail_microversion_id is None:
        raise HandledException(
            f"Could not find the thumbnail tab for {onshape_document["name"]} in the latest version of a document"
        )
    thumbnail_urls = upload_thumbnails(api, thumbnail_path, thumbnail_microversion_id)

    valid_elements = list(get_valid_elements(contents))

    valid_element_ids = {onshape_element["id"] for onshape_element in valid_elements}

    elements_to_reload = list(
        get_elements_to_reload(document_ref, valid_elements, reload_context)
    )

    save_element_operations = [
        asyncio.to_thread(
            save_element,
            api,
            document_ref,
            version_path,
            onshape_element,
            reload_context,
        )
        for onshape_element in elements_to_reload
    ]

    await asyncio.gather(*save_element_operations)

    # Collect list of element ids in same order as the Onshape Tab manager
    ordered_ids = [
        element_id
        for element_id in get_ordered_element_ids(contents)
        if element_id in valid_element_ids
    ]

    # Delete any elements present in the current order but not the new order
    elements_to_delete = set(document_ref.elements.keys()) - set(ordered_ids)
    for element_id in elements_to_delete:
        document_ref.elements.element(element_id).delete()
        document_ref.configurations.configuration(element_id).delete()

    preserved_document = reload_context.get_document(document_id)
    # Document order is externally managed, so just set the document directly
    document_ref.set(
        Document(
            name=onshape_document["name"],
            thumbnailUrls=thumbnail_urls,
            instanceId=version_path.instance_id,
            elementOrder=ordered_ids,
            sortAlphabetically=preserved_document.sortAlphabetically,
        ),
    )
    return len(elements_to_reload)


def build_reload_context(library_ref: LibraryRef, reload_all: bool) -> ReloadContext:
    reload_context = ReloadContext(reload_all=reload_all)
    for document_ref in library_ref.documents.list():
        reload_context.save_document(document_ref.id, document_ref.get())

        for element in document_ref.elements.list():
            element_id = element.id
            reload_context.save_element(element_id, element.get())

    return reload_context


async def reload_document(
    api: Api,
    documents_ref: DocumentsRef,
    document_path: DocumentPath,
    reload_context: ReloadContext,
) -> int:
    latest_version_path = await asyncio.to_thread(
        get_latest_version_path, api, document_path
    )

    document_ref = documents_ref.document(document_path.document_id)
    if not document_ref.maybe_get():
        # Document doesn't exist, create it immediately
        return await save_document(
            api, document_ref, latest_version_path, reload_context
        )

    if not reload_context.should_reload_document(latest_version_path):
        return 0

    # Refresh document
    return await save_document(api, document_ref, latest_version_path, reload_context)


@router.post("/reload-documents" + connect.library_route())
@require_access_level()
async def reload_documents(**kwargs):
    """Saves the contents of the latest versions of all documents managed by FRC Design Lib into the database."""
    api = connect.get_api()
    library_ref = connect.get_library_ref()

    reload_all = connect.get_query_bool("reloadAll", False)

    reload_context = build_reload_context(library_ref, reload_all)

    count = 0

    documents_ref = library_ref.documents

    operations = []
    for document_id in documents_ref.keys():
        document_path = DocumentPath(document_id)
        operations.append(
            reload_document(api, documents_ref, document_path, reload_context)
        )

    results = await asyncio.gather(*operations)
    count = sum(results)

    clean_favorites(library_ref)

    return {"savedElements": count}


def clean_favorites(library_ref: LibraryRef) -> None:
    """Removes any favorites in the library that are no longer valid."""

    elements: dict[str, Element] = {}
    for document_ref in library_ref.documents.list():
        for element_ref in document_ref.elements.list():
            elements[element_ref.id] = element_ref.get()

    for user_data_ref in library_ref.user_data.list():
        for favorite in user_data_ref.favorites.list():
            element = elements.get(favorite.id)
            if element == None:
                continue

            # We have to remove all invisible favorites
            # This is necessary to prevent issues with reordering, as, e.g., Move to top is ambiguous with hidden elements
            if element.isVisible == True:
                continue

            user_data_ref.favorites.remove(favorite.id)


@router.post("/set-visibility" + connect.library_route())
@require_access_level()
def set_visibility(**kwargs):
    """Sets the visibility of one or more elements in a document."""
    library_ref = connect.get_library_ref()
    document_id = connect.get_body_arg("documentId")
    element_ids = connect.get_body_arg("elementIds")
    is_visible = connect.get_body_arg("isVisible")

    document_ref = library_ref.documents.document(document_id)

    if not is_visible:
        for user_data_ref in library_ref.user_data.list():
            for favorite in user_data_ref.favorites.list():
                if favorite.id in element_ids:
                    user_data_ref.favorites.remove(favorite.id)

    for element_id in element_ids:
        document_ref.elements.element(element_id).update({"isVisible": is_visible})
    return {"success": True}


@router.post("/set-document-sort" + connect.library_route())
@require_access_level()
def set_document_sort(**kwargs):
    library_ref = connect.get_library_ref()
    document_id = connect.get_body_arg("documentId")
    sort_alphabetically = connect.get_body_arg("sortAlphabetically")

    library_ref.documents.child(document_id).update(
        {"sortAlphabetically": sort_alphabetically}
    )
    return {"success": True}


@router.post("/document-order" + connect.library_route())
@require_access_level()
def set_document_order(**kwargs):
    library_ref = connect.get_library_ref()
    new_document_order = connect.get_body_arg("documentOrder")

    library_ref.documents.set_order(new_document_order)
    return {"success": True}


@router.post("/document" + connect.library_route())
@require_access_level()
async def add_document_route(**kwargs):
    api = connect.get_api()
    library_ref = connect.get_library_ref()

    new_document_id = connect.get_body_arg("newDocumentId")
    new_path = DocumentPath(new_document_id)
    selected_document_id = connect.get_optional_body_arg("selectedDocumentId")

    try:
        document_name = get_document(api, new_path)["name"]
    except:
        raise HandledException("Failed to find the specified document.")

    try:
        latest_version = get_latest_version_path(api, new_path)
    except:
        raise HandledException("Failed to find a document version to use.")

    document_order = library_ref.documents.keys()
    if new_document_id in document_order:
        raise HandledException("Document has already been added to library.")

    if selected_document_id:
        if selected_document_id not in document_order:
            raise HandledException("Selected document not found in library.")

        selected_index = document_order.index(selected_document_id)
        document_order.insert(selected_index + 1, new_document_id)
    else:
        document_order.append(new_document_id)

    document_ref = library_ref.documents.document(new_document_id)
    await save_document(api, document_ref, latest_version, ReloadContext())

    # Update order after we've successfully added the document
    library_ref.documents.set_order(document_order)
    return {"name": document_name}


@router.delete("/document" + connect.library_route())
@require_access_level()
def delete_document(**kwargs):
    library_ref = connect.get_library_ref()

    document_id = connect.get_query_param("documentId")
    library_ref.documents.remove(document_id)

    clean_favorites(library_ref)
    return {"Success": True}

File: old/src/api/documents.ts
/**
 * Admin server functions for reloading and managing the library document/element database.
 */
import { createServerFn } from "@tanstack/react-start";
import { FieldValue, Timestamp } from "firebase-admin/firestore";
import {
    Library,
    Vendor,
    getVendorName,
    ElementType
} from "../api-utils/client-models";
import { OAuthApi } from "../onshape-api/client/oauth-api";
import { AccessLevel, getAccessLevel } from "../onshape-api/endpoints/users";
import { getLatestVersion } from "../onshape-api/endpoints/versions";
import { getDocument, getContents } from "../onshape-api/endpoints/documents";
import { getFeatures } from "../onshape-api/endpoints/part-studios";
import { getAssembly } from "../onshape-api/endpoints/assemblies";
import { getConfiguration } from "../onshape-api/endpoints/configurations";
import { ElementPath, DocumentPath, InstancePath } from "../onshape-api/path";
import { getLibrary, LibraryRef } from "../connect/library";
import { getOnshapeApi } from "../routes/auth/-auth.server";
import {
    DocumentData,
    ElementData,
    FastenInfo,
    MateLocation,
    VersionInfo
} from "../connect/db-models";
import { uploadThumbnails, uploadDocumentThumbnails } from "../connect/storage";
import { parseOnshapeConfiguration } from "../configurations/parse-configuration";
import {
    ConfigurationParameterType,
    ConfigurationResult
} from "../configurations/configuration-models";

const LATEST_DOCUMENT_SCHEMA = 1;
const LATEST_ELEMENT_SCHEMA = 1;

async function getAppAccessLevel(onshapeApi: OAuthApi): Promise<AccessLevel> {
    const override = process.env.ACCESS_LEVEL_OVERRIDE;
    if (override) return override as AccessLevel;

    const adminTeam = process.env.ADMIN_TEAM;
    if (!adminTeam)
        throw new Error(
            "ADMIN_TEAM or ACCESS_LEVEL_OVERRIDE must be configured"
        );

    return getAccessLevel(onshapeApi, adminTeam);
}

async function requireAccess(
    onshapeApi: OAuthApi,
    required: AccessLevel = AccessLevel.EDITOR
): Promise<void> {
    const level = await getAppAccessLevel(onshapeApi);
    if (
        required === AccessLevel.EDITOR &&
        (level === AccessLevel.EDITOR || level === AccessLevel.ADMIN)
    )
        return;
    if (required === AccessLevel.ADMIN && level === AccessLevel.ADMIN) return;
    throw new Error("Insufficient permissions");
}

// ─── Vendor parsing ──────────────────────────────────────────────────────────

function parseNameVendor(name: string): Vendor | undefined {
    const words = name.toUpperCase().match(/\b(\w+)\b/g) ?? [];
    for (const word of words) {
        const vendor = Object.values(Vendor).find(
            (v) => (v as string).toUpperCase() === word
        );
        if (vendor !== undefined) return vendor;
    }
    return undefined;
}

function parseVendors(
    name: string,
    configuration?: ConfigurationResult
): Vendor[] {
    const nameVendor = parseNameVendor(name);
    if (nameVendor) return [nameVendor];

    if (!configuration) return [];

    const vendors = new Set<Vendor>();
    for (const param of configuration.parameters) {
        if (param.type !== ConfigurationParameterType.ENUM) continue;
        for (const option of param.options) {
            const vendor = parseNameVendor(option.name);
            if (vendor) {
                vendors.add(vendor);
                continue;
            }
            const byFullName = Object.values(Vendor).find(
                (v) =>
                    getVendorName(v).toUpperCase() === option.name.toUpperCase()
            );
            if (byFullName) vendors.add(byFullName);
        }
    }
    return [...vendors];
}

interface SavedElement {
    elementSchema?: number;
    isVisible: boolean;
    isOpenComposite: boolean;
    fastenInfo?: FastenInfo;
    microversionId?: string;
}

interface SavedDocument {
    documentSchema?: number;
    sortAlphabetically: boolean;
    instanceId?: string;
}

class ReloadContext {
    private readonly elements: Map<string, SavedElement> = new Map();
    private readonly documents: Map<string, SavedDocument> = new Map();

    constructor(public readonly reloadAll: boolean = false) {}

    saveElement(elementId: string, data: Partial<ElementData>): void {
        this.elements.set(elementId, {
            elementSchema: data.elementSchema,
            isVisible: data.isVisible ?? false,
            isOpenComposite: data.isOpenComposite ?? false,
            fastenInfo: data.fastenInfo,
            microversionId: data.microversionId
        });
    }

    getElement(elementId: string): SavedElement {
        return (
            this.elements.get(elementId) ?? {
                isVisible: false,
                isOpenComposite: false
            }
        );
    }

    shouldReloadElement(elementId: string, microversionId: string): boolean {
        if (this.reloadAll) return true;
        const saved = this.getElement(elementId);
        if (!saved.elementSchema || saved.elementSchema < LATEST_ELEMENT_SCHEMA)
            return true;
        return saved.microversionId !== microversionId;
    }

    saveDocument(documentId: string, data: Partial<DocumentData>): void {
        this.documents.set(documentId, {
            documentSchema: data.documentSchema,
            sortAlphabetically: data.sortAlphabetically ?? true,
            instanceId: data.instanceId
        });
    }

    getDocument(documentId: string): SavedDocument {
        return this.documents.get(documentId) ?? { sortAlphabetically: true };
    }

    shouldReloadDocument(instancePath: InstancePath): boolean {
        if (this.reloadAll) return true;
        const saved = this.getDocument(instancePath.documentId);
        if (
            !saved.documentSchema ||
            saved.documentSchema < LATEST_DOCUMENT_SCHEMA
        )
            return true;
        return saved.instanceId !== instancePath.instanceId;
    }
}

function searchFeatures(
    features: any[],
    mateLocation: MateLocation,
    path: string[] = []
): FastenInfo | undefined {
    for (const feature of features) {
        if (feature.featureType === "mateConnector") {
            const fullPath = [...path, ...feature.featureData.occurrence];
            return {
                mateConnectorId: feature.id,
                mateLocation,
                path: fullPath
            };
        }
    }
    return undefined;
}

function parseFastenInfoFromPartStudio(featureList: any): FastenInfo {
    for (const feature of featureList.features) {
        if (feature.featureType === "mateConnector") {
            return {
                mateConnectorId: feature.featureId,
                mateLocation: MateLocation.FEATURE,
                path: []
            };
        }
    }
    throw new Error("Failed to find a valid Mate connector feature.");
}

function parseFastenInfoFromAssembly(assemblyInfo: any): FastenInfo {
    const rootAssembly = assemblyInfo.rootAssembly;

    const fromFeatures = searchFeatures(
        rootAssembly.features,
        MateLocation.FEATURE
    );
    if (fromFeatures) return fromFeatures;

    const parts: any[] = assemblyInfo.parts;
    const subAssemblies: any[] = assemblyInfo.subAssemblies;
    let partCounter = 0;
    let subAssemblyCounter = 0;

    for (const instance of rootAssembly.instances) {
        const path = [instance.id as string];
        if (instance.type === "Part") {
            const part = parts[partCounter++];
            const mateConnectors: any[] = part.mateConnectors ?? [];
            if (mateConnectors.length > 0) {
                return {
                    mateConnectorId: mateConnectors[0].featureId,
                    path,
                    mateLocation: MateLocation.PART
                };
            }
        } else if (instance.type === "Assembly") {
            const subAssembly = subAssemblies[subAssemblyCounter++];
            const found = searchFeatures(
                subAssembly.features,
                MateLocation.SUBASSEMBLY,
                path
            );
            if (found) return found;
        }
    }

    throw new Error(
        "Failed to find a valid Mate connector feature or instance with a Mate connector."
    );
}

async function parseFastenInfo(
    onshapeApi: OAuthApi,
    elementPath: ElementPath,
    elementType: ElementType
): Promise<FastenInfo> {
    if (elementType === ElementType.PART_STUDIO) {
        const featureList = await getFeatures(onshapeApi, elementPath);
        return parseFastenInfoFromPartStudio(featureList);
    } else {
        const assemblyInfo = await getAssembly(onshapeApi, elementPath, {
            includeMateConnectors: true,
            includeMateFeatures: true
        });
        return parseFastenInfoFromAssembly(assemblyInfo);
    }
}

// ─── Element & document saving ───────────────────────────────────────────────

const VALID_ELEMENT_TYPES = new Set<string>([
    ElementType.ASSEMBLY,
    ElementType.PART_STUDIO
]);

function getValidElements(contents: any): any[] {
    return (contents.elements as any[]).filter((e) =>
        VALID_ELEMENT_TYPES.has(e.elementType)
    );
}

const ENTRY_TYPE_GROUP = "BTElementGroup-1458";
const ENTRY_TYPE_ELEMENT = "BTDocumentElementReference-2484";

function* traverseEntry(entry: any): Generator<string> {
    if (entry.btType === ENTRY_TYPE_GROUP) {
        for (const child of entry.groups) yield* traverseEntry(child);
    } else if (entry.btType === ENTRY_TYPE_ELEMENT) {
        yield entry.elementId;
    }
}

function getOrderedElementIds(contents: any): string[] {
    const ids: string[] = [];
    for (const entry of contents.folders.groups) {
        ids.push(...traverseEntry(entry));
    }
    return ids;
}

function parseVersion(
    versionDict: any,
    documentId: string
): { instancePath: InstancePath; versionInfo: VersionInfo } {
    return {
        instancePath: {
            documentId,
            instanceId: versionDict.id,
            instanceType: "v"
        },
        versionInfo: {
            name: versionDict.name,
            createdAt: Timestamp.fromDate(new Date(versionDict.createdAt))
        }
    };
}

async function saveElement(
    onshapeApi: OAuthApi,
    libraryRef: LibraryRef,
    documentId: string,
    versionPath: InstancePath,
    onshapeElement: any,
    reloadContext: ReloadContext
): Promise<void> {
    const elementType: ElementType = onshapeElement.elementType;
    const elementId: string = onshapeElement.id;
    const microversionId: string = onshapeElement.microversionId;

    const elementPath: ElementPath = { ...versionPath, elementId };
    const docRef = libraryRef.documents.document(documentId);

    const onshapeConfig = await getConfiguration(onshapeApi, elementPath);
    let configuration: ConfigurationResult | undefined;
    let configurationId: string | undefined;

    if (onshapeConfig.configurationParameters.length > 0) {
        configuration = parseOnshapeConfiguration(onshapeConfig);
        await docRef.configurations.configuration(elementId).set(configuration);
        configurationId = elementId;
    }

    const thumbnailUrls = await uploadThumbnails(
        onshapeApi,
        elementPath,
        microversionId
    );

    const preserved = reloadContext.getElement(elementId);

    let fastenInfo: FastenInfo | undefined;
    if (preserved.fastenInfo) {
        fastenInfo = await parseFastenInfo(onshapeApi, elementPath, elementType);
    }

    await docRef.elements.element(elementId).set({
        elementSchema: LATEST_ELEMENT_SCHEMA,
        name: onshapeElement.name,
        vendors: parseVendors(onshapeElement.name, configuration),
        elementType,
        documentId,
        instanceId: versionPath.instanceId,
        microversionId,
        isVisible: preserved.isVisible,
        isOpenComposite: preserved.isOpenComposite,
        fastenInfo,
        configurationId,
        thumbnailUrls
    } satisfies ElementData);
}

async function saveDocument(
    onshapeApi: OAuthApi,
    libraryRef: LibraryRef,
    documentId: string,
    versionPath: InstancePath,
    versionInfo: VersionInfo,
    reloadContext: ReloadContext
): Promise<number> {
    const documentRef = libraryRef.documents.document(documentId);

    const [onshapeDocument, contents] = await Promise.all([
        getDocument(onshapeApi, versionPath),
        getContents(onshapeApi, versionPath)
    ]);

    const thumbnailUrls = await uploadDocumentThumbnails(
        onshapeApi,
        onshapeDocument,
        contents,
        versionPath
    );

    const validElements = getValidElements(contents);
    const validElementIds = new Set(
        validElements.map((e: any) => e.id as string)
    );

    const elementsToReload = validElements.filter((e: any) =>
        reloadContext.shouldReloadElement(e.id, e.microversionId)
    );

    await Promise.all(
        elementsToReload.map((e: any) =>
            saveElement(
                onshapeApi,
                libraryRef,
                documentId,
                versionPath,
                e,
                reloadContext
            )
        )
    );

    const orderedIds = getOrderedElementIds(contents).filter((id) =>
        validElementIds.has(id)
    );

    const existingIds = await documentRef.elements.keys();
    const toDelete = existingIds.filter((id) => !validElementIds.has(id));
    await Promise.all(
        toDelete.map(async (id) => {
            await documentRef.elements.element(id).delete();
            await documentRef.configurations.configuration(id).delete();
        })
    );

    const preserved = reloadContext.getDocument(documentId);
    await documentRef.set({
        documentSchema: LATEST_DOCUMENT_SCHEMA,
        name: onshapeDocument.name,
        thumbnailUrls,
        instanceId: versionPath.instanceId,
        elementOrder: orderedIds,
        sortAlphabetically: preserved.sortAlphabetically,
        versionInfo
    } satisfies DocumentData);

    return elementsToReload.length;
}

async function reloadDocument(
    onshapeApi: OAuthApi,
    libraryRef: LibraryRef,
    documentId: string,
    reloadContext: ReloadContext
): Promise<number> {
    const versionDict = await getLatestVersion(onshapeApi, { documentId });
    const { instancePath, versionInfo } = parseVersion(versionDict, documentId);

    const documentRef = libraryRef.documents.document(documentId);
    const existing = await documentRef.get();

    if (existing && !reloadContext.shouldReloadDocument(instancePath)) {
        return 0;
    }

    return saveDocument(
        onshapeApi,
        libraryRef,
        documentId,
        instancePath,
        versionInfo,
        reloadContext
    );
}

async function buildReloadContext(
    libraryRef: LibraryRef,
    reloadAll: boolean
): Promise<ReloadContext> {
    const context = new ReloadContext(reloadAll);
    const documentIds = await libraryRef.documents.keys();

    await Promise.all(
        documentIds.map(async (documentId) => {
            const docRef = libraryRef.documents.document(documentId);
            const docData = await docRef.get();
            if (docData) context.saveDocument(documentId, docData);

            const elementEntries = await docRef.elements.list();
            for (const { id, data } of elementEntries) {
                context.saveElement(id, data);
            }
        })
    );

    return context;
}

async function cleanFavorites(libraryRef: LibraryRef): Promise<void> {
    const elements = new Map<string, ElementData>();
    const documentIds = await libraryRef.documents.keys();

    await Promise.all(
        documentIds.map(async (documentId) => {
            const docRef = libraryRef.documents.document(documentId);
            const elementEntries = await docRef.elements.list();
            for (const { id, data } of elementEntries) {
                elements.set(id, data);
            }
        })
    );

    const userDataEntries = await libraryRef.userData.list();
    await Promise.all(
        userDataEntries.map(async ({ id: userId }) => {
            const userRef = libraryRef.userData.userDataFor(userId);
            const favorites = await userRef.favorites.list();
            await Promise.all(
                favorites.map(async ({ id: favoriteId }) => {
                    const element = elements.get(favoriteId);
                    if (!element || !element.isVisible) {
                        await userRef.favorites.remove(favoriteId);
                    }
                })
            );
        })
    );
}

// ─── Server functions ────────────────────────────────────────────────────────

/**
 * Reloads the contents of all documents in the library from Onshape.
 * Requires MEMBER-level access.
 */
export const reloadDocuments = createServerFn()
    .inputValidator((data: { library: Library; reloadAll?: boolean }) => data)
    .handler(async ({ data: { library, reloadAll = false } }) => {
        const onshapeApi = await getOnshapeApi();
        await requireAccess(onshapeApi);

        const libraryRef = getLibrary(library);
        const reloadContext = await buildReloadContext(libraryRef, reloadAll);
        const documentIds = await libraryRef.documents.keys();

        const results = await Promise.all(
            documentIds.map((documentId) =>
                reloadDocument(onshapeApi, libraryRef, documentId, reloadContext)
            )
        );

        await cleanFavorites(libraryRef);

        return { savedElements: results.reduce((a, b) => a + b, 0) };
    });

/**
 * Sets the visibility of one or more elements in a document.
 * Requires MEMBER-level access.
 */
export const setElementVisibility = createServerFn()
    .inputValidator(
        (data: {
            library: Library;
            documentId: string;
            elementIds: string[];
            isVisible: boolean;
        }) => data
    )
    .handler(
        async ({ data: { library, documentId, elementIds, isVisible } }) => {
            const onshapeApi = await getOnshapeApi();
            await requireAccess(onshapeApi);

            const libraryRef = getLibrary(library);
            const documentRef = libraryRef.documents.document(documentId);

            if (!isVisible) {
                const userDataEntries = await libraryRef.userData.list();
                await Promise.all(
                    userDataEntries.map(async ({ id: userId }) => {
                        const userRef = libraryRef.userData.userDataFor(userId);
                        const favorites = await userRef.favorites.list();
                        await Promise.all(
                            favorites
                                .filter(({ id }) => elementIds.includes(id))
                                .map(({ id }) => userRef.favorites.remove(id))
                        );
                    })
                );
            }

            await Promise.all(
                elementIds.map((elementId) =>
                    documentRef.elements
                        .element(elementId)
                        .update({ isVisible })
                )
            );

            return { success: true };
        }
    );

/**
 * Sets whether a document should be sorted alphabetically.
 * Requires MEMBER-level access.
 */
export const setSortAlphabetically = createServerFn()
    .inputValidator(
        (data: {
            library: Library;
            documentId: string;
            sortAlphabetically: boolean;
        }) => data
    )
    .handler(async ({ data: { library, documentId, sortAlphabetically } }) => {
        const onshapeApi = await getOnshapeApi();
        await requireAccess(onshapeApi);

        await getLibrary(library)
            .documents.document(documentId)
            .update({ sortAlphabetically });
        return { success: true };
    });

/**
 * Sets the order of documents in the library.
 * Requires MEMBER-level access.
 */
export const setDocumentOrder = createServerFn()
    .inputValidator(
        (data: { library: Library; documentOrder: string[] }) => data
    )
    .handler(async ({ data: { library, documentOrder } }) => {
        const onshapeApi = await getOnshapeApi();
        await requireAccess(onshapeApi);

        await getLibrary(library).ref.update({ documentOrder });
        return { success: true };
    });

/**
 * Adds a new document to the library and loads its latest version.
 * Requires MEMBER-level access.
 */
export const addDocument = createServerFn()
    .inputValidator(
        (data: {
            library: Library;
            documentId: string;
            selectedDocumentId?: string;
        }) => data
    )
    .handler(
        async ({
            data: { library, documentId: newDocumentId, selectedDocumentId }
        }) => {
            const onshapeApi = await getOnshapeApi();
            await requireAccess(onshapeApi);

            const documentPath: DocumentPath = { documentId: newDocumentId };

            let documentName: string;
            try {
                const doc = await getDocument(onshapeApi, documentPath);
                documentName = doc.name;
            } catch {
                throw new Error("Failed to find the specified document.");
            }

            let versionDict: any;
            try {
                versionDict = await getLatestVersion(onshapeApi, documentPath);
            } catch {
                throw new Error("Failed to find a document version to use.");
            }

            const libraryRef = getLibrary(library);
            const documentOrder = await libraryRef.documents.keys();

            if (documentOrder.includes(newDocumentId)) {
                throw new Error("Document has already been added to library.");
            }

            if (selectedDocumentId) {
                const selectedIndex = documentOrder.indexOf(selectedDocumentId);
                if (selectedIndex === -1)
                    throw new Error("Selected document not found in library.");
                documentOrder.splice(selectedIndex + 1, 0, newDocumentId);
            } else {
                documentOrder.push(newDocumentId);
            }

            const { instancePath, versionInfo } = parseVersion(
                versionDict,
                newDocumentId
            );
            await saveDocument(
                onshapeApi,
                libraryRef,
                newDocumentId,
                instancePath,
                versionInfo,
                new ReloadContext()
            );

            // Update order after successfully saving the document
            await libraryRef.ref.update({ documentOrder });
            return { name: documentName };
        }
    );

/**
 * Removes a document and all its elements from the library.
 * Requires MEMBER-level access.
 */
export const deleteDocument = createServerFn()
    .inputValidator((data: { library: Library; documentId: string }) => data)
    .handler(async ({ data: { library, documentId } }) => {
        const onshapeApi = await getOnshapeApi();
        await requireAccess(onshapeApi);

        const libraryRef = getLibrary(library);
        const documentRef = libraryRef.documents.document(documentId);

        const [elementEntries, configEntries] = await Promise.all([
            documentRef.elements.list(),
            documentRef.configurations.list()
        ]);

        await Promise.all([
            ...elementEntries.map(({ id }) =>
                documentRef.elements.element(id).delete()
            ),
            ...configEntries.map(({ id }) =>
                documentRef.configurations.configuration(id).delete()
            )
        ]);

        await documentRef.delete();
        await libraryRef.ref.update({
            documentOrder: FieldValue.arrayRemove(documentId)
        });

        await cleanFavorites(libraryRef);
        return { success: true };
    });
